src/BTree.mo

```diff
@@ -0,0 +1,110 @@
+/// Imperative sequences as B-Trees.
+
+import A "Array";
+import I "Iter";
+import List "List";
+import Option "Option";
+import Order "Order";
+import P "Prelude";
+import Prim "mo:⛔";
+
+module {
+
+  /// Constants we use to shape the tree.
+  /// See https://en.wikipedia.org/wiki/B-tree#Definition
+  module Constants {
+    let MAX_CHILDREN = 4;
+  };
+
+  public type Compare<K> = {
+    compare : (K, K) -> Order.Order
+  };
+
+  public type Data<K, V> = [(K, V)];
+
+  public type Index<K, V> = {
+    data : Data<K, V>;
+    trees : [Tree<K, V>];
+  };
+
+  public type Tree<K, V> = {
+    #index : Index<K, V>;
+    #data : Data<K, V>;
+  };
+
+  func find_data<K, V>(data : Data<K, V>, find_k : K, c : Compare<K>) : ?V {
+    for ((k, v) in data.vals()) {
+      if (c.compare(k, find_k) == #equal) { return ?v };
+    };
+    return null
+  };
+
+  func find<K, V>(t : Tree<K, V>, k : K, c : Compare<K>) : ?V {
+    switch t {
+      case (#data(d)) { return find_data<K, V>(d, k, c) };
+      case (#index(i)) {
+        for (j in I.range(0, i.data.size())) {
+          switch (c.compare(k, i.data[j].0)) {
+            case (#equal) { return ?i.data[j].1 };
+            case (#less) { return find<K, V>(i.trees[j], k, c) };
+            case _ { }
+          }
+        };
+        find<K, V>(i.trees[i.data.size()], k, c)
+      };
+    };
+  };
+
+  /// Check that a B-Tree instance observes invariants of B-Trees.
+  /// Invariants ensure performance is what we expect.
+  /// For testing and debugging.
+  public module Check {
+
+    type CompareOp<K> = {
+      compare : (?K, ?K) -> Order.Order
+    };
+
+    func compareOp<K>(c : Compare<K>) : CompareOp<K> = {
+      compare = func (k1 : ?K, k2 : ?K) : Order.Order {
+        switch (k1, k2) {
+        case (null, null) { assert false; loop {} };
+        case (null, _) #less;
+        case (_, null) #greater;
+        case (?k1, ?k2) c.compare(k1, k2)
+        }
+      }
+    };
+
+    public func check<K, V>(c : Compare<K>, t : Tree<K, V>) {
+      rec(null, compareOp(c), t, null)
+    };
+
+    func rec<K, V>(lower : ?K, c : CompareOp<K>, t : Tree<K, V>, upper : ?K) {
+      switch t {
+        case (#data(d)) { data(lower, c, d, upper) };
+        case (#index(i)) { index(lower, c, i, upper) };
+      }
+    };
+
+    func data<K, V>(lower : ?K, c : CompareOp<K>, d : Data<K, V>, upper : ?K) {
+      var prev_k : ?K = null;
+      for ((k, _) in d.vals()) {
+        assert (c.compare(prev_k, ?k) != #greater);
+        assert (c.compare(lower, ?k) != #greater);
+        assert (c.compare(?k, upper) != #greater);
```

Comment:
In a single data node BTree (which is valid) or in the very first call to `data()` in the `index()` method (line 101), it looks like the `check` method calls 
```
rec(null, compareOp(c), t, null)
```
which then makes its way down to call 
```
data<K, V>(lower : ?K, c : CompareOp<K>, d : Data<K, V>, upper : ?K)
```
with an upper key of `null`, which then hits the 
```
  case (_, null) #greater;
```
case of `compareOp` and should fail when it hits the 
```
assert (c.compare(?k, upper) != #greater);
```
case of this `data` method

Reply:
Yep, when I wrote the code, I had a feeling it was wrong, but it's helping the right code become more apparent. :). Thanks for taking this close look, and writing this out.

To fix things here, I suspect that I need to distinguish positive inf and negative inf, rather than encoding *both* of them with `null`.

BTW: This fix is also related to your other question about the relationship between an inner's node's keys and its subtrees (there is no key to lower bound the first subtree, and no key to upper bound the last one, so you need some special values that are always smaller/bigger than everything else).

Reply:
Update: All of this mess is fixed now.